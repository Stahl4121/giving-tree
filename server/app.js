// Stripe Server

const app = require('express')();
const bodyParser = require('body-parser');
const stripe = require('stripe')('sk_test_avCwuzIvg45JxkjItTyqRGH600JEoryzyP');
const admin = require('firebase-admin');
const serviceAccount = require("./secret/givingtree-cfs-firebase-adminsdk-grrpd-764871e22c.json");

admin.initializeApp({
  credential: admin.credential.cert(serviceAccount),
  databaseURL: "https://givingtree-cfs.firebaseio.com"
});

app.use(bodyParser.text());
app.use(bodyParser.urlencoded({ extended: true }));

app.post('/charge', async (req, res) => {
  let source = req.body.split(' amount: ')[0];
  let amount = req.body.split(' amount: ')[1].split(' description: ')[0];
  let description = req.body.split(' amount: ')[1].split(' description: ')[1].split(' account: ')[0];
  let account = req.body.split(' amount: ')[1].split(' description: ')[1].split(' account: ')[1];
  try {
    let { status } = await stripe.charges.create({
      amount: amount,
      currency: 'usd',
      description: description,
      source: source,
    }, {
      stripeAccount: account,
    }).then(function (charge) {
      // asynchronously called
      return charge;
    });

    res.json({ status });
  } catch (err) {
    console.log(err);
    res.status(500).end();
  }
});

app.post('/create', async (req, res) => {
  try {
    let connected_account_id = await stripe.oauth.token({
      grant_type: 'authorization_code',
      code: req.body,
    }).then(function (response) {
      // asynchronously called
      return response.stripe_user_id;
    });
    return res.json({ 'stripeId': connected_account_id });

  } catch (err) {
    res.status(500).end;
  }
})



//Get current user's ID token
//If metaadmin, call "getUserByEmail" w/ the community foundation email, update status to accepted/rejected
//If not metaadmin, get current user's ID token and set custom claim 'cf' to true
app.post('/setCustomClaims', (req, res) => {

  // Get the ID token passed.
  const idToken = req.body.idToken;
  const cfEmail = req.body.cfEmail;
  const cfStatus = req.body.cfStatus;

  // Verify the ID token and decode its payload.
  admin.auth().verifyIdToken(idToken).then((claims) => {
    console.log('verified token\n');

    //Check if metaadmin
    if (typeof claims.admin !== 'undefined' && claims.admin) {
      //Check that cfEmail and cfStatus were provided in request
      if (typeof cfEmail !== 'undefined' && typeof cfStatus !== 'undefined'
        && cfEmail !== '' && cfStatus !== '' && typeof claims.cfId !== 'undefined' ) {
        //Get the CF User and update their permissions
        admin.auth().getUserByEmail(cfEmail).then((user) => {
          if (cfStatus === 'current') {
            admin.auth().setCustomUserClaims(user.uid, { cfId: claims.cfId, status: 'current' })
              .then(function () {
                // Tell client to refresh token on user.
                res.end(JSON.stringify({ status: 'success' }));
              })
              .catch((error) => {
                res.end(JSON.stringify({ status: error }));
              });
          }
          else if (cfStatus === 'denied') {
              admin.auth().setCustomUserClaims(user.uid, { cfId: claims.cfId, status: 'denied' })
                .then(function () {
                  // Tell client to refresh token on user.
                  res.end(JSON.stringify({ status: 'success' }));
                })
                .catch((error) => {
                  res.end(JSON.stringify({ status: error }));
                });
          }
        })
          .catch((error) => {
            res.end(JSON.stringify({ status: error }));
          });
      }
    }
    else {
      console.log('not metaadmin\n');
      //Not metaadmin, so we this post must be made from CF Account Request
      //Get and store the cf document record id in the claims object
      admin.auth().getUser(claims.sub)
        .then((user) => {
          console.log('got user\n');
          admin.firestore().collection('communityFoundations').where('personal_email', '==', user.email)
            .get()
            .then(function (querySnapshot) {
              querySnapshot.forEach(function (doc) {
              console.log('got cf!\n' );
                admin.auth().setCustomUserClaims(claims.sub, { cfId: doc.id, status: 'requested' })
                  .then(() => {
                  console.log('set cc!\n');
                    // Tell client to refresh token on user.
                    res.end(JSON.stringify({ status: 'success' }));
                  })
                  .catch((error) => {
                  console.log('failed cc!\n');

                    res.end(JSON.stringify({ status: error }));
                  });
                  console.log('after cc\n');
              });
            })
            .catch((error) => {
              res.end(JSON.stringify({ status: error }));
            });
        })
        .catch((error) => {
          res.end(JSON.stringify({ status: error }));
        });
        console.log('end of not metaadmin\n');
    }
  })
  .catch((error) => {
    console.log('NOOO at the end\n');
    res.end(JSON.stringify({ status: error }));
  });
});

app.listen(process.env.PORT);
// app.listen(9000, () => console.log('Listening on port 9000')); 	