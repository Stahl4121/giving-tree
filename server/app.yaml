#TODO: Figure out what else should go in here
runtime: nodejs10
