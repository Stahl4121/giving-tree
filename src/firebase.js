//Reasoning for this file: https://stackoverflow.com/questions/48492047/where-do-i-initialize-firebase-app-in-react-application

// Firebase App (the core Firebase SDK) is always required and
// must be listed before other Firebase SDKs
import * as firebase from "firebase/app";

// Add the Firebase services that you want to use
import "firebase/auth";
import "firebase/firestore";

//Import everything, not recommended for production apps
//import firebase from "firebase";

const firebaseConfig = {
  apiKey: "AIzaSyAkeCK6w-qbvHzaOdEyg_PtpGQbElAv0_M",
  authDomain: "givingtree-cfs.firebaseapp.com",
  databaseURL: "https://givingtree-cfs.firebaseio.com",
  projectId: "givingtree-cfs",
  storageBucket: "givingtree-cfs.appspot.com",
  messagingSenderId: "709659064913",
  appId: "1:709659064913:web:d9decf22beb442f2e79670",
  measurementId: "G-KZ6PNG9K64"
};


firebase.initializeApp(firebaseConfig);

export default !firebase.apps.length ? firebase.initializeApp(firebaseConfig) : firebase.app();