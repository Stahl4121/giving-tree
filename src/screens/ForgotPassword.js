import React, { Component } from 'react';
import { withRouter } from 'react-router-dom';

import firebase from '../firebase.js';
import * as helper from '../helpers/ValidationHelper.js';
import Snack from '../components/Snack.js';

import { withStyles } from '@material-ui/core/styles';
import Container from '@material-ui/core/Container';
import Avatar from '@material-ui/core/Avatar';
import TextField from "@material-ui/core/TextField";
import Button from "@material-ui/core/Button";
import LockOutlinedIcon from '@material-ui/icons/LockOutlined';
import Grid from "@material-ui/core/Grid";
import Link from "@material-ui/core/Link";
import Typography from "@material-ui/core/Typography";


const styles = theme => ({
  root: {
    marginTop: theme.spacing(8),
    display: 'flex',
    flexDirection: 'column',
    alignItems: 'center',
  },
  avatar: {
    margin: theme.spacing(1),
    backgroundColor: theme.palette.secondary.main,
  },
  form: {
    marginTop: theme.spacing(1),
  },
  submit: {
    margin: theme.spacing(3, 0, 2),
  },
  links: {
    textDecoration: 'none'
  },
  errorMsg: {
    color: 'red',
    marginTop: theme.spacing(1),
  }
});

function ForgotPassword(props) {
  const { classes } = props;

  return (
    <Container component="main" maxWidth="xs">
      <div className={classes.root}>
        <Avatar className={classes.avatar}>
          <LockOutlinedIcon />
        </Avatar>
        <Typography component="h1" variant="h5" className={classes.header}>
          Reset Password
                    </Typography>
        <ForgotPasswordForm classes={classes} />
        <SignInSignUpLink classes={classes} />
      </div>
    </Container>
  );
}

const INITIAL_STATE = {
  email: '',
  errors: {
    email: '',
    submit: '',
  },
  isValid: false,
  success: false,
};

export class ForgotPasswordFormBase extends Component {
  constructor(props) {
    super(props);
    this.state = { ...INITIAL_STATE };
    this.validateForm = this.validateForm.bind(this);
    this.onChange = this.onChange.bind(this);
    this.onSubmit = this.onSubmit.bind(this);
  }
  
  onSubmit(event, fb=firebase) {
    if (this.state.isValid) {
      const { email } = this.state;
      fb.auth().sendPasswordResetEmail(email)
        .then(() => {
          // Email sent.
          this.setState({ ...INITIAL_STATE, success: true });
        }).catch((error) => {
          // An error happened.
          this.setState({ errors: { ...this.state.errors, submit: error.message } });
        });
    }

    event.preventDefault();
  }

  //Written generically to match SignUp/AccountRequest/SignIn, even though there's only one field 
  onChange(event) {
    const { name, value } = event.target;
    this.setState({ [name]: value });

    this.setState({ errors: { ...this.state.errors, submit: '', [name]: helper.validateField(name, value) } },
      this.validateForm
    );
  };

  validateForm() {
    const noEmptyFields = Object.values(this.state).filter((s) => { return s === '' }).length === 0
    const noErrors = Object.values(this.state.errors).filter((s) => { return s !== '' }).length === 0
    this.setState({ isValid: noErrors && noEmptyFields });
  }

  render() {
    const {
      email,
      errors,
      isValid,
      success,
    } = this.state;
    
    const { classes } = this.props;
    return (
      <form className={classes.form} onSubmit={this.onSubmit} noValidate>
        {success &&
          <Snack message='Success! A password reset email has been sent.' />
        }
        <Typography component="h6" className={classes.errorMsg} >
          {errors.submit}
        </Typography>
        <TextField
          variant="outlined"
          margin="normal"
          required
          fullWidth
          name="email"
          value={email}
          onChange={this.onChange}
          type="text"
          label="Email Address"
          error={errors.email !== ""}
          helperText={errors.email}
          autoFocus
        />
        <Button className={classes.submit}
          type="submit"
          fullWidth
          variant="contained"
          color="primary"
          disabled={!isValid}
        >
          Reset Password
                        </Button>
      </form>
    );
  }
}
const SignInSignUpLink = ({ classes }) => (
  <Grid container>
    <Grid item xs>
      <Link href='/signin' variant="body2">
        Ready to sign in?
            </Link>
    </Grid>
    <Grid item>
      <Link href='/signup' variant="body2">
        Don't have an account? Sign Up
            </Link>
    </Grid>
  </Grid>
);

const ForgotPasswordForm = withRouter(ForgotPasswordFormBase);

export default withStyles(styles)(ForgotPassword);

export { ForgotPasswordForm, SignInSignUpLink };