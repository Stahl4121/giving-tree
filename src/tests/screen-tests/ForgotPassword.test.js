import React from 'react';
import { shallow, mount } from 'enzyme';
import AuthUserContext from '../../auth/context.js';

import {firestore as FIRESTORE} from "firebase/app";

import ForgotPassword from '../../screens/ForgotPassword.js';
import { ForgotPasswordFormBase, SignInSignUpLink } from '../../screens/ForgotPassword.js';
import { StaticRouter } from 'react-router-dom';

import { createMount } from '@material-ui/core/test-utils';
import { createMuiTheme } from '@material-ui/core';
import { ThemeProvider } from '@material-ui/core/styles';


function MockTheme({ children }) {
  const theme = createMuiTheme({});
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
}

describe('ForgotPassword', () => {

  let wrap;

  beforeEach(() => {
    wrap = mount(
      <StaticRouter>
        <ForgotPassword/>
      </StaticRouter>
      
    )
  });

  it('Renders correctly', () => {  
    expect(wrap).toMatchSnapshot();
  });

})


describe('ForgotPasswordForm', () => {
  
  let wrap;
  let mount;

  const error = {message: 'It went wrong!'};
  const firebase = {
    auth : function() {
      return {sendPasswordResetEmail: function(email) {
        console.log(email+' got passed');
        return {then: function(callback) {
          callback();
          return {catch: function(callback) {
            callback(error);
            return "catch successful"
          }}
        }}
      }};
    }
  };


  const props = {
    classes: {
      root: "ForgotPassword-root-519",
      avatar: "ForgotPassword-avatar-520",
      form: "ForgotPassword-form-521",
      submit: "ForgotPassword-submit-522",
      links: "ForgotPassword-links-523",
      errorMsg: "ForgotPassword-errorMsg-524"
    }
  }

  const mockedEvent = { target: {name: 'bah', value: 'fug'} }

  beforeEach(() => {
    mount = createMount();
    wrap = shallow(
      <ForgotPasswordFormBase {...props}/>
    );
  });

  afterEach(() => {
    mount.cleanUp();
  });

  it('Updates with onChange correctly', () => {    
    wrap.find("WithStyles(ForwardRef(TextField))").simulate('change', mockedEvent)
    expect(wrap.state('errors')).toEqual({ email: '', submit: '', bah: '' })
  });
  
  it('Updates with validateForm correctly', () => {  
    wrap.find("WithStyles(ForwardRef(TextField))").simulate('change', mockedEvent)
    expect(wrap.state('isValid')).toBeFalsy()
  });

  it('Uses with onSubmit correctly (failure)', () => {  
    const fakeEvent = { preventDefault: () => console.log('preventDefault') };
    wrap.setState({isValid: true});
    const inst = wrap.instance();
    inst.onSubmit(fakeEvent);
    expect(wrap.state('success')).toBeFalsy()
  });

  it('Uses with onSubmit correctly (success)', () => {  
    const fakeEvent = { preventDefault: () => console.log('preventDefault') };
    const inst = wrap.instance();
    wrap.setState({isValid: 'blah@yada.net'});
    wrap.setState({isValid: true});
    inst.onSubmit(fakeEvent, firebase);
    expect(wrap.state('success')).toBeTruthy();
    expect(wrap.state('errors').submit).toEqual('It went wrong!');
  });

});


describe('SigninSignupLink', () => {
  
  let wrap;

  
  beforeEach(() => {
    wrap = shallow(
        <SignInSignUpLink/>
    )
  });

  it('Renders correctly', () => {  
    expect(wrap).toMatchSnapshot();
  });

  
});