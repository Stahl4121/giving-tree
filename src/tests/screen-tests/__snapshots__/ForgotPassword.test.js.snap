// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ForgotPassword Renders correctly 1`] = `ReactWrapper {}`;

exports[`SigninSignupLink Renders correctly 1`] = `ShallowWrapper {}`;
